File: drills/ts/solutions/nc142_lowest_common_ancestor_bst.ts
export class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

export function lowestCommonAncestor(root: TreeNode | null, p: number, q: number): number {
  // The ordering does all the work. If both targets are below the current value
  // go left, if both are above go right, and otherwise this node is the split
  // point -- which is the answer. No searching for either node first, and no
  // comparing of paths.
  let node = root;
  while (node !== null) {
    if (p < node.val && q < node.val) node = node.left;
    else if (p > node.val && q > node.val) node = node.right;
    else return node.val;
  }
  return -1;
}

File: drills/python/solutions/nc142_lowest_common_ancestor_bst.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def lowestCommonAncestor(root, p, q):
    # The ordering does all the work. If both targets are below the current
    # value go left, if both are above go right, and otherwise this node is the
    # split point -- which is the answer. No searching for either node first,
    # and no comparing of paths.
    node = root
    while node is not None:
        if p < node.val and q < node.val:
            node = node.left
        elif p > node.val and q > node.val:
            node = node.right
        else:
            return node.val
    return -1
